### PR TITLE
add E-branchformer to i6_models

### PR DESCRIPTION
**Pull request by Judyxujj**
This PR adds E-branchformer to i6_models

### CODE REVIEW

File: i6_models/parts/e_branchformer/merge.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        kernel_size: kernel size of the depthwise convolution layer
+        dropout: dropout probability
+    """
+
+    input_dim: int
+    kernel_size: int
+    dropout: float
+
+    def __post_init__(self):
+        super().__post_init__()
```
Comment (michelwi) [inline i6_models/parts/e_branchformer/merge.py:27]:
```suggestion
```
If you just call the same function of the parent class, then this code is not necessary. If you remove it, the function is just inherited.

File: i6_models/parts/e_branchformer/cgmlp.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+        # convolutional spatial gating unit (csgu)
+        x_1, x_2 = x.chunk(2, dim=-1)  # (B,T,3*F), (B,T,3*F)
+        x_2 = self.layer_norm_csgu(x_2)
+        # conv layers expect shape [B,F,T] so we have to transpose here
+        x_2 = x_2.transpose(1, 2)  # [B,3*F,T]
+        x_2 = self.depthwise_conv(x_2)
+        x_2 = x_2.transpose(1, 2)  # [B,T,3*F]
+        x = x_1 * x_2
+        x = nn.functional.dropout(x, p=self.dropout, training=self.training)
```
Comment (kuacakuaca) [inline i6_models/parts/e_branchformer/cgmlp.py:76]: is it intended that dropout is applied two times, one after the CSGU block, and one after the linear layer?

Comment (Judyxujj) [reply i6_models/parts/e_branchformer/cgmlp.py:76]: yes, in espnet implementation, they also apply twice dropout

File: i6_models/parts/e_branchformer/merge.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        https://arxiv.org/abs/2210.00077 for more merge module variants
+        """
+        super().__init__()
+
+        self.depthwise_conv = nn.Conv1d(
+            in_channels=model_cfg.input_dim * 2,
+            out_channels=model_cfg.input_dim * 2,
+            kernel_size=model_cfg.kernel_size,
+            padding=(model_cfg.kernel_size - 1) // 2,
+            groups=model_cfg.input_dim * 2 // 2,
```
Comment (kuacakuaca) [inline i6_models/parts/e_branchformer/merge.py:41]: why is here `groups` not the same as `in_channels`?

File: i6_models/assemblies/e_branchformer/e_branchformer_v1.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
+from __future__ import annotations
+
+__all__ = [
+    "EbranchformerBlockV1Config",
+    "EbranchformerBlockV1",
+    "EbranchformerEncoderV1Config",
+    "EbranchformerrEncoderV1",
```
Comment (Atticus1806) [inline i6_models/assemblies/e_branchformer/e_branchformer_v1.py:7]: I think there is a Typo in the whole module, I will just mark here, but there are a few lines affected.

File: i6_models/assemblies/e_branchformer/e_branchformer_v1.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        self.ff_1 = PositionwiseFeedForwardV1(cfg=cfg.ff_cfg)
+        self.mhsa = MHSAV1(cfg=cfg.mhsa_cfg)
+        self.cgmlp = ConvolutionalGatingMLPV1(model_cfg=cfg.cgmlp_cfg)
+        self.merger = MergerV1(model_cfg=cfg.merger_cfg)
+        self.ff_2 = PositionwiseFeedForwardV1(cfg=cfg.ff_cfg)
+        self.final_layer_norm = torch.nn.LayerNorm(cfg.ff_cfg.input_dim)
+
+    def forward(self, x: torch.Tensor, /, sequence_mask: torch.Tensor) -> torch.Tensor:
+        """
+        :param tensor: input tensor of shape [B,T,F]
```
Comment (Atticus1806) [inline i6_models/assemblies/e_branchformer/e_branchformer_v1.py:65]: Different spacing of shape compared to two lines below.

Comment (Atticus1806) [reply i6_models/assemblies/e_branchformer/e_branchformer_v1.py:65]:
```suggestion
        :param tensor: input tensor of shape [B, T, F]
```

File: i6_models/assemblies/e_branchformer/e_branchformer_v1.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    num_layers: int
+
+    # nested configurations
+    frontend: ModuleFactoryV1
+    block_cfg: EbranchformerBlockV1Config
+
+
+class EbranchformerrEncoderV1(nn.Module):
+    """
+    Implementation of the Branchformer with Enhanced merging (short E-branchformer), as in the original publication.
```
Comment (Atticus1806) [inline i6_models/assemblies/e_branchformer/e_branchformer_v1.py:95]: Whats the correct shortform here? Saw e_branchformer, e-branchformer and E-branchformer now. Please unify.

File: i6_models/assemblies/e_branchformer/e_branchformer_v1.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+    The model consists of a frontend and a stack of N e-branchformer blocks.
+    C.f. https://arxiv.org/pdf/2210.00077.pdf
+    """
+
+    def __init__(self, cfg: EbranchformerEncoderV1Config):
+        """
+        :param cfg: conformer encoder configuration with subunits for frontend and conformer blocks
```
Comment (Atticus1806) [inline i6_models/assemblies/e_branchformer/e_branchformer_v1.py:102]:
```suggestion
        :param cfg: E-branchformer encoder configuration with subunits for frontend and conformer blocks
```
(or which other capitalization is the usual)

File: i6_models/assemblies/e_branchformer/e_branchformer_v1.py [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+        """
+        super().__init__()
+
+        self.frontend = cfg.frontend()
+        self.module_list = torch.nn.ModuleList([EbranchformerBlockV1(cfg.block_cfg) for _ in range(cfg.num_layers)])
+
+    def forward(self, data_tensor: torch.Tensor, sequence_mask: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        """
+        :param data_tensor: input tensor of shape [B, T', F]
+        :param sequence_mask: mask tensor where 0 defines positions within the sequence and 1 outside, shape: [B, T']
```
Comment (Atticus1806) [inline i6_models/assemblies/e_branchformer/e_branchformer_v1.py:112]: c.f. potentially #30

File: i6_models/assemblies/e_branchformer/e_branchformer_v1.py [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+        :return: (output, out_seq_mask)
+            where output is torch.Tensor of shape [B, T, F'],
+            out_seq_mask is a torch.Tensor of shape [B, T]
+
+        F: input feature dim, F': internal and output feature dim
```
Comment (Atticus1806) [inline i6_models/assemblies/e_branchformer/e_branchformer_v1.py:117]: for time you use `T'` for the input / unmodified time (which makes sense imo) but I feel like then here `F` and `F'` should also be flipped.

File: i6_models/parts/e_branchformer/cgmlp.py [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+            kernel_size=model_cfg.kernel_size,
+            padding=(model_cfg.kernel_size - 1) // 2,
+            groups=model_cfg.hidden_dim // 2,
+        )
+        self.linear_out = nn.Linear(in_features=model_cfg.hidden_dim // 2, out_features=model_cfg.input_dim, bias=True)
+        self.dropout = model_cfg.dropout
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """
+        :param x: shape [B,T,F], F=input_dim
```
Comment (Atticus1806) [inline i6_models/parts/e_branchformer/cgmlp.py:61]:
```suggestion
        :param x: shape [B, T, F], F=input_dim
```

File: i6_models/parts/e_branchformer/cgmlp.py [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+        :return: shape [B,T,F], F=input_dim
```
Comment (Atticus1806) [inline i6_models/parts/e_branchformer/cgmlp.py:62]:
```suggestion
        :return: shape [B, T, F], F=input_dim
```

File: i6_models/parts/e_branchformer/cgmlp.py [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+        """
+        x = self.layer_norm_input(x)  # (B,T,F)
+        x = self.linear_ff(x)  # (B,T,6*F)
```
Comment (Atticus1806) [inline i6_models/parts/e_branchformer/cgmlp.py:65]:
```suggestion
        x = self.linear_ff(x)  # [B, T, 6*F]
```

File: i6_models/parts/e_branchformer/cgmlp.py [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+        x = self.activation(x)
+
+        # convolutional spatial gating unit (csgu)
+        x_1, x_2 = x.chunk(2, dim=-1)  # (B,T,3*F), (B,T,3*F)
```
Comment (Atticus1806) [inline i6_models/parts/e_branchformer/cgmlp.py:69]:
```suggestion
        x_1, x_2 = x.chunk(2, dim=-1)  # [B, T, 3*F], [B, T, 3*F]
```

Comment (Atticus1806) [reply i6_models/parts/e_branchformer/cgmlp.py:69]: Same as above, this is only valid for matching `hidden_dim`

File: i6_models/parts/e_branchformer/cgmlp.py [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+        x_2 = self.layer_norm_csgu(x_2)
+        # conv layers expect shape [B,F,T] so we have to transpose here
```
Comment (Atticus1806) [inline i6_models/parts/e_branchformer/cgmlp.py:71]:
```suggestion
        # conv layers expect shape [B, F, T] so we have to transpose here
```

File: i6_models/parts/e_branchformer/cgmlp.py [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+        x_2 = x_2.transpose(1, 2)  # [B,3*F,T]
```
Comment (Atticus1806) [inline i6_models/parts/e_branchformer/cgmlp.py:72]: See shape comments above.

File: i6_models/parts/e_branchformer/cgmlp.py [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+        x_2 = self.depthwise_conv(x_2)
+        x_2 = x_2.transpose(1, 2)  # [B,T,3*F]
```
Comment (Atticus1806) [inline i6_models/parts/e_branchformer/cgmlp.py:74]: See shape comments above.

File: i6_models/parts/e_branchformer/cgmlp.py [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+        x = x_1 * x_2
```
Comment (Atticus1806) [inline i6_models/parts/e_branchformer/cgmlp.py:75]: I think here the shape would be nice to see too.

File: i6_models/parts/e_branchformer/merge.py [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+        )
+        self.linear_ff = nn.Linear(in_features=2 * model_cfg.input_dim, out_features=model_cfg.input_dim, bias=True)
+        self.dropout = model_cfg.dropout
+
+    def forward(self, x_1: torch.Tensor, x_2: torch.Tensor) -> torch.Tensor:
+        x_concat = torch.cat([x_1, x_2], dim=-1)  # (B,T,2F)
```
Comment (Atticus1806) [inline i6_models/parts/e_branchformer/merge.py:47]: Consistency of shape docs, here and lines below.

File: tests/test_e_branchformer.py [hunk 19]
```diff
@@ -0,0 +1,36 @@
+from itertools import product
+
+import torch
+from torch import nn
+
+from i6_models.parts.e_branchformer.cgmlp import ConvolutionalGatingMLPV1Config, ConvolutionalGatingMLPV1
+from i6_models.parts.e_branchformer.merge import MergerV1Config, MergerV1
+
+
+def test_ConvolutionalGatingMLPV1():
```
Comment (Atticus1806) [inline tests/test_e_branchformer.py:10]: If we only test shapes maybe this should reflect in the test name. Same below.

**Review by Atticus1806 — CHANGES_REQUESTED** [review]
Looks mostly good. Changes are rather docstring problems instead of conceptional problems.

File: i6_models/parts/e_branchformer/merge.py (hunk 3 above)
Comment (mmz33) [reply i6_models/parts/e_branchformer/merge.py:41]: Yes this should be same as `in_channels`. See also here for reference: https://github.com/espnet/espnet/blob/master/espnet2/asr/encoder/e_branchformer_encoder.py#L102C32-L102C32

**Review by mmz33 — CHANGES_REQUESTED** [review]

**Review by Atticus1806 — APPROVED** [review]
Looks good, thanks for your work.

File: i6_models/parts/e_branchformer/cgmlp.py [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+
+    def __init__(self, model_cfg: ConvolutionalGatingMLPV1Config):
+        super().__init__()
+
+        self.layer_norm_input = nn.LayerNorm(model_cfg.input_dim)
+        self.linear_ff = nn.Linear(in_features=model_cfg.input_dim, out_features=model_cfg.hidden_dim, bias=True)
+        self.activation = model_cfg.activation
+        self.layer_norm_csgu = nn.LayerNorm(model_cfg.hidden_dim // 2)
+        self.depthwise_conv = nn.Conv1d(
+            in_channels=model_cfg.hidden_dim // 2,
```
Comment (JackTemaki) [inline i6_models/parts/e_branchformer/cgmlp.py:50]: You could add a check to the `check_valid()` verifying that the hidden dim is divisible by 2.

File: i6_models/parts/e_branchformer/merge.py [hunk 21]
```diff
@@ -0,0 +1,55 @@
...
+class MergerV1Config(ModelConfiguration):
+    """
+    Attributes:
+        input_dim: input dimension
+        kernel_size: kernel size of the depthwise convolution layer
+        dropout: dropout probability
+    """
+
+    input_dim: int
+    kernel_size: int
```
Comment (JackTemaki) [inline i6_models/parts/e_branchformer/merge.py:23]: Are we allowed to have even kernel_size here?

**Review by JackTemaki — COMMENTED** [review]
Just two small things, I did not verify the actual logic though.

File: i6_models/parts/e_branchformer/cgmlp.py (hunk 20 above)
Comment (Judyxujj) [reply i6_models/parts/e_branchformer/cgmlp.py:51]: `assert self.hidden_dim % 2 == 0, "ConvolutionalGatingMLPV1 only supports even hidden_dim"`
I have the `check_valid()` in `ConvolutionalGatingMLPV1Config`

File: i6_models/parts/e_branchformer/merge.py (hunk 21 above)
Comment (Judyxujj) [reply i6_models/parts/e_branchformer/merge.py:23]: I think even `kernel_size` is not forbidden, but usually an odd kernel_size is used for the convolution so I follow the valid check as for the other convolution in i6_model for standardization. Do you suggest that we remove the check here?

File: i6_models/parts/e_branchformer/merge.py [hunk 22]
```diff
@@ -0,0 +1,62 @@
...
+    def __post__init__(self):
+        super().__post_init__()
+        self.check_valid()
+
+
+class MergerV1(nn.Module):
+    def __init__(self, model_cfg: MergerV1Config):
+        """
+        The merge module to merge the outputs of local extractor and global extractor
+        Here we take the best variant from the E-branchformer paper, refer to
```
Comment (christophmluscher) [inline i6_models/parts/e_branchformer/merge.py:38]:
```suggestion
        Here we take the best variant from the E-branchformer paper (Fig. 3c), refer to
```

Comment (christophmluscher) [reply i6_models/parts/e_branchformer/merge.py:38]: I would suggest this small change to make it easier for the user.

**Review by christophmluscher — APPROVED** [review]
One small doc comment

**Comment by christophmluscher** [comment]
@mmz33 could you please review? then we can merge :)

**Review by mmz33 — APPROVED** [review]
